src/OrbitBase/include/OrbitBase/JoinFutures.h

```diff
@@ -6,13 +6,88 @@
 #define ORBIT_BASE_JOIN_FUTURES_H_
 
 #include "OrbitBase/Future.h"
+#include "OrbitBase/FutureHelpers.h"
+#include "OrbitBase/Promise.h"
 #include "absl/types/span.h"
 
+namespace orbit_base_internal {
+
+template <typename T>
+class SharedStateJoin {
+  orbit_base::Promise<std::vector<T>> promise;
+  absl::Mutex mutex;
+  int incompleted_futures;
+  std::vector<std::optional<T>> results;
+
+  void SetResultsInPromise() {
+    std::vector<T> output;
+    output.reserve(results.size());
+    std::transform(results.begin(), results.end(), std::back_inserter(output),
+                   [](std::optional<T>& element) {
+                     CHECK(element.has_value());
+                     return *std::move(element);
+                   });
+    results.clear();
+    promise.SetResult(std::move(output));
+  }
+
+ public:
+  void SetNumberOfFutures(size_t size) {
+    absl::MutexLock lock{&mutex};
+    incompleted_futures = static_cast<int>(size);
+    results.resize(size);
+  }
+
+  void SetResult(int index, const T& argument) {
+    absl::MutexLock lock{&mutex};
+    results[index] = argument;
```

Comment:
To make sure the same index cant be set multiple times
```suggestion
    absl::MutexLock lock{&mutex};
    CHECK(!results[index].has_value());
    results[index] = argument;
```

Reply:
Done. But gracefully.